core/payments/payments.go

```diff
@@ -268,39 +271,82 @@ func (s *service) GetSubscriptionStatus(ctx context.Context, req *pb.RpcMembersh
 
 	out := convertMembershipStatus(status)
 
-	// 5. Save to cache. Lifetime - min(subscription ends, now + TTL)
+	// 4 - Save to cache. Lifetime - min(subscription ends, now + TTL)
 	// update only status, not tiers
 	err = s.cache.CacheSet(&out, nil)
 	if err != nil {
 		log.Warn("can not save subscription status to cache", zap.Error(err))
 		// return nil, ErrCacheProblem
 	}
 
-	isDiffTier := (cachedStatus != nil) && (cachedStatus.Data != nil) && (cachedStatus.Data.Tier != status.Tier)
-	isDiffStatus := (cachedStatus != nil) && (cachedStatus.Data != nil) && (cachedStatus.Data.Status != model.MembershipStatus(status.Status))
-	isEmailDiff := (cachedStatus != nil) && (cachedStatus.Data != nil) && (cachedStatus.Data.UserEmail != status.UserEmail)
+	log.Debug("subscription status", zap.Any("from server", status), zap.Any("cached", cachedStatus))
 
-	log.Debug("subscription status", zap.Any("from server", status), zap.Any("cached", cachedStatus), zap.Bool("isEmailDiff", isEmailDiff))
+	// 5 - Send all messages to the client if needed
+	if !isUpdateRequired(cacheErr, cachedStatus, status) {
+		// no need to send events or enable/disable cache
+		return &out, nil
+	}
+	s.updateStatus(ctx, status)
 
-	// 6. If tier or status has changed -> send event
-	if cachedStatus != nil && !isDiffTier && !isDiffStatus && !isEmailDiff {
+	// 6 - Enable or disable cache (only if status has changed)
+	// also automatically will enable cash in N minutes after DisableForNextMinutes() call
+	s.enableCache(status)
+	s.disableCache(status)
+
+	return &out, nil
+}
+
+func canReturnCachedStatus(s *pb.RpcMembershipGetStatusResponse) bool {
+	return s != nil && s.Data != nil && (s.Error == nil || s.Error.Code == pb.RpcMembershipGetStatusResponseError_NULL)
+}
+
+func isUpdateRequired(cacheErr error, cachedStatus *pb.RpcMembershipGetStatusResponse, status *proto.GetSubscriptionResponse) bool {
+	// 1 - If cache was empty or expire or disabled or NoCache flag was passed
+	// -> treat at is if data was different
+	isCacheEmpty := (cacheErr != nil) && !errors.Is(cacheErr, cache.ErrCacheDisabled)
+
+	if isCacheEmpty {
+		log.Debug("subscription status treated as changed because cache was empty/expired/NoCache flag was passed")
+		return true
+	}
+
+	// 2 - Check if tier or status has changed
+	isDiffTier := false
+	isDiffStatus := false
+	isEmailDiff := false
+
+	if canReturnCachedStatus(cachedStatus) {
```

Comment:
It seems we have some biz logic problem here. 
If **canReturnCachedStatus** returns false (e.g. cached status is nil or contains some error), then **isUpdateRequired** returns false. Bit imho in this case status should be updated in cache as well